File: DesignPatterns/01_GOF23/DP_examples/C/03_Behavioral/Mediator/mediator.c
#include "mediator.h"

// 静态实例
static PrivateChatRoom* privateChatRoomInstance = NULL;
static PublicChatRoom* publicChatRoomInstance = NULL;

// 用户方法实现
static void User_SendMessage(User* self, IMediator* room, User* desUser, const char* message) {
    room->ShowMessage(room, self, desUser, message);
}

static void User_Destroy(User* self) {
    if (self->name) {
        free(self->name);
    }
    free(self);
}

// 创建用户
User* CreateUser(const char* name) {
    User* user = (User*)malloc(sizeof(User));
    if (user) {
        user->name = strdup(name);
        user->SendMessage = User_SendMessage;
        user->Destroy = User_Destroy;
    }
    return user;
}

// 私聊房间方法实现
static void PrivateChatRoom_ShowMessage(IMediator* self, User* oriUser, User* desUser, const char* message) {
    printf("[PrivateChatRoom][%s] sent to [%s]: %s\n", oriUser->name, desUser->name, message);
}

static void PrivateChatRoom_Destroy(IMediator* self) {
    // 单例模式，不在这里释放
}

// 获取私聊房间实例
PrivateChatRoom* GetPrivateChatRoomInstance() {
    if (!privateChatRoomInstance) {
        privateChatRoomInstance = (PrivateChatRoom*)malloc(sizeof(PrivateChatRoom));
        if (privateChatRoomInstance) {
            privateChatRoomInstance->mediator.ShowMessage = PrivateChatRoom_ShowMessage;
            privateChatRoomInstance->mediator.Destroy = PrivateChatRoom_Destroy;
        }
    }
    return privateChatRoomInstance;
}

// 公共聊天室方法实现
static void PublicChatRoom_ShowMessage(IMediator* self, User* user, User* desUser, const char* message) {
    printf("[PublicChatRoom][%s]: %s\n", user->name, message);
}

static void PublicChatRoom_Destroy(IMediator* self) {
    // 单例模式，不在这里释放
}

static void PublicChatRoom_ShowMessagePublic(PublicChatRoom* self, User* user, const char* message) {
    self->mediator.ShowMessage((IMediator*)self, user, NULL, message);
}

// 获取公共聊天室实例
PublicChatRoom* GetPublicChatRoomInstance() {
    if (!publicChatRoomInstance) {
        publicChatRoomInstance = (PublicChatRoom*)malloc(sizeof(PublicChatRoom));
        if (publicChatRoomInstance) {
            publicChatRoomInstance->mediator.ShowMessage = PublicChatRoom_ShowMessage;
            publicChatRoomInstance->mediator.Destroy = PublicChatRoom_Destroy;
            publicChatRoomInstance->ShowMessagePublic = PublicChatRoom_ShowMessagePublic;
        }
    }
    return publicChatRoomInstance;
}

File: DesignPatterns/01_GOF23/DP_examples/Lua/03_Behavioral/Mediator/Mediator.lua
-- 中介者接口
local IMediator = {}
IMediator.__index = IMediator

function IMediator:new()
    local o = {}
    setmetatable(o, self)
    return o
end

function IMediator:ShowMessage(oriUser, desUser, message)
    error("Subclass must implement ShowMessage method")
end

-- 用户类
local User = {}
User.__index = User

function User:new(name)
    local o = {name = name}
    setmetatable(o, self)
    return o
end

function User:GetName()
    return self.name
end

function User:SendMessage(room, desUser, message)
    room:ShowMessage(self, desUser, message)
end

-- 私聊房间
local PrivateChatRoom = {}
PrivateChatRoom.__index = PrivateChatRoom
setmetatable(PrivateChatRoom, IMediator)

-- 静态实例
local privateChatRoomInstance = nil

function PrivateChatRoom:new()
    local o = IMediator:new()
    setmetatable(o, self)
    return o
end

function PrivateChatRoom:ShowMessage(oriUser, desUser, message)
    print(string.format("[PrivateChatRoom][%s] sent to [%s]: %s", oriUser:GetName(), desUser:GetName(), message))
end

-- 获取私聊房间实例
function GetPrivateChatRoomInstance()
    if not privateChatRoomInstance then
        privateChatRoomInstance = PrivateChatRoom:new()
    end
    return privateChatRoomInstance
end

-- 公共聊天室
local PublicChatRoom = {}
PublicChatRoom.__index = PublicChatRoom
setmetatable(PublicChatRoom, IMediator)

-- 静态实例
local publicChatRoomInstance = nil

function PublicChatRoom:new()
    local o = IMediator:new()
    setmetatable(o, self)
    return o
end

function PublicChatRoom:ShowMessage(user, desUser, message)
    print(string.format("[PublicChatRoom][%s]: %s", user:GetName(), message))
end

function PublicChatRoom:ShowMessagePublic(user, message)
    self:ShowMessage(user, nil, message)
end

-- 获取公共聊天室实例
function GetPublicChatRoomInstance()
    if not publicChatRoomInstance then
        publicChatRoomInstance = PublicChatRoom:new()
    end
    return publicChatRoomInstance
end

return { User = User, GetPrivateChatRoomInstance = GetPrivateChatRoomInstance, GetPublicChatRoomInstance = GetPublicChatRoomInstance }
